engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java

```diff
@@ -573,6 +612,46 @@ mul self (other : V) = V.Val self.a*other.a
     assertEquals(42, fourtyTwoAsV.getMember("a").asInt());
   }
 
+  @Test
+  public void selfTypeConversion() throws Exception {
+    final URI uri = new URI("memory://self_type_conversion.enso");
+    final Source src =
+        Source.newBuilder(
+                "enso",
+                """
+    from Standard.Base import all
+    type My_Type
+        Value x
+        f self y = self.x+y
+
+    type Convertible_Type
+        A x
+
+    type Inconvertible_Type
+        B x
+
+    My_Type.from (that : Convertible_Type) = My_Type.Value that.x+1
+
+    static_convertible = My_Type.f (Convertible_Type.A 23) 1000
+    static_inconvertible = My_Type.f (Inconvertible_Type.B 23) 1000
+    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = ctx.eval(src);
+    var convertible = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "static_convertible");
+    assertEquals(1024, convertible.asInt());
+
+    try {
+      var invalid_static_call =
+          module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "static_inconvertible");
+      fail("Expecting an exception, not: " + invalid_static_call);
+    } catch (PolyglotException e) {
+      assertTypeError("`self`", "My_Type", "Inconvertible_Type", e.getMessage());
+    }
+  }
```

Comment:
This test also ensures that not only values of requested type but also values _convertible_ to it are supported, if they go through the conversion.